java-checks/src/main/java/org/sonar/java/checks/security/AndroidMobileDatabaseEncryptionKeysCheck.java

```diff
@@ -0,0 +1,121 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.security;
+
+import javax.annotation.CheckForNull;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.ExpressionsHelper;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import static org.sonar.java.checks.helpers.ExpressionsHelper.getSingleWriteUsage;
+
+@Rule(key = "S6301")
+public class AndroidMobileDatabaseEncryptionKeysCheck extends AbstractMethodDetection {
+
+  private static final String JAVA_LANG_STRING = "java.lang.String";
+
+  private static final MethodMatchers SQLITE_DATABASE_CONSTRUCTOR = MethodMatchers.create()
+    .ofTypes("net.sqlcipher.database.SQLiteDatabase")
+    .constructor()
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers SQLITE_DATABASE_METHODS = MethodMatchers.create()
+    .ofTypes("net.sqlcipher.database.SQLiteDatabase")
+    .names("changePassword", "openDatabase", "openOrCreateDatabase", "create")
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers REALM_CONFIGURATION_BUILDER_ENCRYPTION_KEY = MethodMatchers.create()
+    .ofTypes("io.realm.RealmConfiguration$Builder")
+    .names("encryptionKey")
+    .addParametersMatcher("byte[]")
+    .build();
+
+  private static final MethodMatchers STRING_TO_CHAR_GET_BYTES = MethodMatchers.create()
+    .ofTypes(JAVA_LANG_STRING)
+    .names("toCharArray", "getBytes")
+    .withAnyParameters()
+    .build();
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return MethodMatchers.or(
+      SQLITE_DATABASE_CONSTRUCTOR,
+      SQLITE_DATABASE_METHODS,
+      REALM_CONFIGURATION_BUILDER_ENCRYPTION_KEY);
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    Arguments arguments = mit.arguments();
+    String messageArg = REALM_CONFIGURATION_BUILDER_ENCRYPTION_KEY.matches(mit) ? "encryptionKey" : "password";
+    ExpressionTree passwordArg = arguments.size() == 1 ? arguments.get(0) : arguments.get(1);
```

Comment:
You are using `withAnyParameters()` in the matchers, it means that you have no guarantee that the argument list is not empty here. Even if there is no method without arguments currently, I would rather explicitly make sure we do not crash, by making sure the argument list is not empty.

Note: if you think specifying the types of every parameter is overkill, you can use `MethodMatchers.ANY`:
```
.addParametersMatcher(MethodMatchers.ANY)
.addParametersMatcher(MethodMatchers.ANY, MethodMatchers.ANY)
```
Alternatively, you can also play with the method taking a predicate to make sure the list is not empty:
```
.addParametersMatcher(l -> !l.isEmpty())
```